goserver/smartonfhir/smartonfhirtest/smartonfhirtest.go

```diff
@@ -0,0 +1,70 @@
+package smartonfhirtest
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"strings"
+	"testing"
+)
+
+// StartFHIRServer starts and returns a server that handles requests sent on
+// the given configPath and returns the given authBaseURL and tokenURL as JSON
+// in the response body.
+func StartFHIRServer(configPath, authBaseURL, tokenURL string) *httptest.Server {
+	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Path == configPath {
+			w.WriteHeader(http.StatusOK)
+			w.Write([]byte(fmt.Sprintf("{\"authorization_endpoint\": \"%s\", \"token_endpoint\": \"%s\"}", authBaseURL, tokenURL)))
+			return
+		}
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}))
+}
+
+// StartFHIRTokenServer starts and returns a server that handles token exchange requests and
+// returns the given token as JSON in the response body. The handler will return status 400 if
+// the request contains form values that do not match the given ones.
+func StartFHIRTokenServer(code, redirectURI, clientID, token string) *httptest.Server {
+	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.FormValue("grant_type") != "authorization_code" || r.FormValue("code") != code || r.FormValue("redirect_uri") != redirectURI || r.FormValue("client_id") != clientID {
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+
+		w.Header()["Content-Type"] = []string{"application/json"}
+		w.WriteHeader(http.StatusOK)
+		w.Write([]byte(fmt.Sprintf("{\"access_token\": \"%s\"}", token)))
+	}))
+}
+
+// ValidateAuthURL validate an authURL by making sure its host and query parameters match the
+// given ones.
+func ValidateAuthURL(t *testing.T, authURL *url.URL, host, clientID, redirectURL, launchID, state, aud string, scopes []string) {
```

Comment:
It will be used by https://github.com/google/meet-on-fhir/blob/fhir-auth-endpoints/goserver/server/server_test.go in a following PR.

Reply:
Makes sense. Ok in that case a few suggestions:
 1) we shouldn't pass testing.T into these methods, prefer just returning an error string and having the actual test call `t.Errorf()`. This makes the code more re-usable in the future.
  2) Could we structure this slightly differently: instead of this method could we create the expected url.URL value and use the cmp package to test that actual == expected? If cmp doesn't end up working due to private fields or something you could also define a custom comparison on url.URL instead of this method (e.g. `CompareAuthURLs(actual, expected *url.URL) (string, bool)`). This way you don't have this super long parameter list and it's easier to extend in the future if you have to verify other parts of the URL.

Reply:
CompareAuthURLs(actual, expected *url.URL) (string, bool) seems to be a generic function to just check the equivalence of two URLs. I'd prefer to have a function signatured AuthURL(host, clientID, redirectURL, launchID, state, aud string, scopes []string) *url.URL which is used by the test to construct URLs and do comparison. WDYT?